File: packages/shared/src/execution-policies.ts
/** DTOs API — políticas de ejecución (ADR-010 P5). */

import type {
  ExecutionMode,
  ExecutionPolicyV1,
  KernelEntityOrigin,
} from './platform-kernel.js';
import type { SignalKind } from './signal-events.js';
import type { AlertChannelType } from './signal-alerts-api.js';
import type { ScanHitDto } from './scan-api.js';

export interface ExecutionPolicySummaryDto {
  id: string;
  name: string;
  mode: ExecutionMode;
  accountId?: string | null;
  strategyDefinitionId?: string | null;
  signalKinds: SignalKind[];
  requireValidatedBacktest: boolean;
  enabled: boolean;
  updatedAt: string;
  createdAt: string;
}

export interface ExecutionPolicyDetailDto extends ExecutionPolicySummaryDto {
  definition: ExecutionPolicyV1;
}

export interface CreateExecutionPolicyDto {
  name: string;
  mode: ExecutionMode;
  accountId?: string | null;
  strategyDefinitionId?: string | null;
  signalKinds?: SignalKind[];
  channels?: AlertChannelType[];
  webhookUrl?: string | null;
  emailTo?: string | null;
  requireValidatedBacktest?: boolean;
  origin?: KernelEntityOrigin;
  enabled?: boolean;
}

export interface UpdateExecutionPolicyDto {
  name?: string;
  mode?: ExecutionMode;
  accountId?: string | null;
  strategyDefinitionId?: string | null;
  signalKinds?: SignalKind[];
  channels?: AlertChannelType[];
  webhookUrl?: string | null;
  emailTo?: string | null;
  requireValidatedBacktest?: boolean;
  enabled?: boolean;
}

export interface ExecutionPolicyResponseDto {
  data: ExecutionPolicyDetailDto;
}

export interface ExecutionPoliciesListResponseDto {
  data: ExecutionPolicySummaryDto[];
}

export type ExecutionActionStatus =
  | 'inform_only'
  | 'alert_dispatched'
  | 'trade_executed'
  | 'skipped'
  | 'live_dry_run_pass'
  | 'live_dry_run_veto';

export interface ExecutionActionResultDto {
  instrumentId: string;
  signalKind: SignalKind;
  status: ExecutionActionStatus;
  reason?: string | null;
  transactionId?: string | null;
  dispatches?: { channel: string; ok: boolean; error?: string | null }[];
}

export interface RouteSignalsRequestDto {
  hits: ScanHitDto[];
}

export interface RouteSignalsResponseDto {
  data: {
    policyId: string;
    mode: ExecutionMode;
    actions: ExecutionActionResultDto[];
  };
}

export interface ExecuteScanJobRequestDto {
  policyId: string;
}

File: apps/api-python/src/bolsa_api/schemas/execution_policies.py
"""DTOs HTTP de políticas de ejecución."""

from pydantic import BaseModel, ConfigDict, Field


class ExecutionPolicySummaryDto(BaseModel):
    model_config = ConfigDict(populate_by_name=True, ser_json_by_alias=True)

    id: str
    name: str
    mode: str
    account_id: str | None = Field(default=None, alias="accountId")
    strategy_definition_id: str | None = Field(default=None, alias="strategyDefinitionId")
    signal_kinds: list[str] = Field(alias="signalKinds")
    require_validated_backtest: bool = Field(alias="requireValidatedBacktest")
    enabled: bool
    updated_at: str = Field(alias="updatedAt")
    created_at: str = Field(alias="createdAt")


class ExecutionPolicyDetailDto(ExecutionPolicySummaryDto):
    definition: dict


class ExecutionPoliciesListResponseDto(BaseModel):
    data: list[ExecutionPolicySummaryDto]


class ExecutionPolicyResponseDto(BaseModel):
    data: ExecutionPolicyDetailDto


class CreateExecutionPolicyRequestDto(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    name: str
    mode: str
    account_id: str | None = Field(default=None, alias="accountId")
    strategy_definition_id: str | None = Field(default=None, alias="strategyDefinitionId")
    signal_kinds: list[str] | None = Field(default=None, alias="signalKinds")
    channels: list[str] | None = None
    webhook_url: str | None = Field(default=None, alias="webhookUrl")
    email_to: str | None = Field(default=None, alias="emailTo")
    require_validated_backtest: bool = Field(default=False, alias="requireValidatedBacktest")
    origin: str = "manual"
    enabled: bool = True


class UpdateExecutionPolicyRequestDto(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    name: str | None = None
    mode: str | None = None
    account_id: str | None = Field(default=None, alias="accountId")
    strategy_definition_id: str | None = Field(default=None, alias="strategyDefinitionId")
    signal_kinds: list[str] | None = Field(default=None, alias="signalKinds")
    channels: list[str] | None = None
    webhook_url: str | None = Field(default=None, alias="webhookUrl")
    email_to: str | None = Field(default=None, alias="emailTo")
    require_validated_backtest: bool | None = Field(default=None, alias="requireValidatedBacktest")
    enabled: bool | None = None


class ExecutionActionResultDto(BaseModel):
    model_config = ConfigDict(populate_by_name=True, ser_json_by_alias=True)

    instrument_id: str = Field(alias="instrumentId")
    signal_kind: str = Field(alias="signalKind")
    status: str
    reason: str | None = None
    transaction_id: str | None = Field(default=None, alias="transactionId")
    dispatches: list[dict] | None = None


class RouteSignalsRequestDto(BaseModel):
    hits: list[dict]


class RouteSignalsResponseDto(BaseModel):
    data: dict


class ExecuteScanJobRequestDto(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    policy_id: str = Field(alias="policyId")
